# Linear Regression


### Theoretical Overview<br>
$$ H(x) = Wx + b $$
$$ cost(W, b) = \frac{1}{m} \sum^m_{i=1} \left( H(x^{(i)}) - y^{(i)} \right)^2 $$<br>
$H(x)$: 주어진 $x$ 값에 대해 예측을 어떻게 할 것인가<br>
$cost(W, b)$: $H(x)$ 가 $y$ 를 얼마나 잘 예측했는가


### Imports

In [6]:
import torch

### Data

In [3]:
#Training Dataset 생성
x_train = torch.FloatTensor([[1], [2], [3]])
y_train = torch.FloatTensor([[2], [4], [6]])
#x_train = torch.FloatTensor([1., 2., 3.])
#y_train = torch.FloatTensor([2., 4., 6.])

### Hypothesis

In [12]:
#W와 b 초기화 
#zeros function = zeros(at::IntArrayRef size, const at::TensorOptions&options={})
# requires_grad는 해당 Tensor에서 행해지는 모든 연산에 대한 미분값 계산 유무이다.
W = torch.zeros(1, requires_grad=True)
b = torch.zeros(1, requires_grad=True)
hypothesis = x_train*W + b

### Cost

In [13]:
#Cost func 직접 구현
cost = torch.mean((hypothesis - y_train)**2)
print(cost)

tensor(18.6667, grad_fn=<MeanBackward0>)


### Gradient descent

In [14]:
#optim은 pytorch에서 제공하는 다양한 optimization algorithm package이다.
#[,] 내에는 학습시킬 tenser를 넣는다. lr 는 learning rate를 의미.
#zero_grad() -> gradient 초기화 // backward() -> gradient 계선 // step() -> gradient 개선

#W에 Gradient값 저장
optimizer = torch.optim.SGD([W, b], lr=0.01)

nb_epochs = 1000
for epoch in range(1, nb_epochs + 1):
    hypothesis = x_train * W + b
    cost = torch.mean((hypothesis - y_train) ** 2)
    
    #optimizer의 gradient 0으로 초기화
    optimizer.zero_grad()
    #costfunc을 미분하여 grad를 채움
    cost.backward()
    #저장된 gradient 값으로 gradient descent 실핼
    optimizer.step()

In [15]:
print(cost)
print(hypothesis)

tensor(0.0006, grad_fn=<MeanBackward0>)
tensor([[2.0373],
        [4.0080],
        [5.9787]], grad_fn=<AddBackward0>)


1000회 반복 시 tensor([[2.0373],[4.0080],[5.9787]], grad_fn=<AddBackward0>)
약 4000회 반복 시 tensor([[2.0000], [4.0000],[6.0000]], grad_fn=<AddBackward0>)